server/src/main/java/org/elasticsearch/index/shard/GlobalCheckpointListeners.java

```diff
@@ -214,16 +220,15 @@ private void notifyListeners(final long globalCheckpoint, final IndexShardClosed
             listeners.clear();
         }
         if (listenersToNotify.isEmpty() == false) {
-            executor.execute(() ->
-                    listenersToNotify
-                            .forEach((listener, t) -> {
-                                /*
-                                 * We do not want to interrupt any timeouts that fired, these will detect that the listener has been
-                                 * notified and not trigger the timeout.
-                                 */
-                                FutureUtils.cancel(t.v2());
-                                notifyListener(listener, globalCheckpoint, e);
-                            }));
+            listenersToNotify
+                .forEach((listener, t) -> {
+                    /*
+                     * We do not want to interrupt any timeouts that fired, these will detect that the listener has been notified and not
+                     * trigger the timeout.
+                     */
+                    FutureUtils.cancel(t.v2());
+                    listener.executor().execute(() -> notifyListener(listener, globalCheckpoint, e));
```

Comment:
Maybe move the wrapping `listener.executor().execute(() ...` into `notifyListener` to dry things up? We only ever invoke `notifyListener` from `listener.executor().execute(() -> ...` anyway.

Reply:
I pushed b565820a862107f74bfdd23a99158a820e1231f9. I mistakenly combined your two suggestions into a single commit, sorry.